include/flamegpu/simulation/detail/MPIEnsemble.h

```diff
@@ -0,0 +1,108 @@
+    enum EnvelopeTag : int {
+        // Sent from worker to manager to request a job index to process
+        RequestJob = 0,
+        // Sent from manager to worker to assign a job index to process in response to AssignJob
+        AssignJob = 1,
+        // Sent from worker to manager to report an error during job execution
+        // If fail fast is enabled, following RequestJob will receive an exit job id (>=plans.size())
+        ReportError = 2,
+        // Sent from worker to manager to report GPUs for telemetry
+        TelemetryDevices = 3,
+    };
+
+ public:
+    const int world_rank;
+    const int world_size;
+    const unsigned int total_runs;
+    /**
+     * Construct the object for managing MPI comms during an ensemble
+     *
+     * Initialises the MPI_Datatype MPI_ERROR_DETAIL, detects world rank and size.
+     *
+     * @param _config The parent ensemble's config (mostly used to check error/verbosity levels)
+     * @param _total_runs The total number of runs to be executed (only used for printing error warnings)
+     */
+    explicit MPIEnsemble(const CUDAEnsemble::EnsembleConfig &_config, unsigned int _total_runs);
+    /**
+     * If world_rank==0, receive any waiting errors and add their details to err_detail
+     * @param err_detail The map to store new error details within
+     * @return The number of errors that occurred.
+     */
+    int receiveErrors(std::multimap<int, AbstractSimRunner::ErrorDetail> &err_detail);
+    /**
+     * If world_rank==0, receive and process any waiting job requests
+     * @param next_run A reference to the int which tracks the progress through the run plan vector
+     * @return The number of runners that have been told to exit (if next_run>total_runs)
+     */
+    int receiveJobRequests(unsigned int &next_run);
+    /**
+     * If world_rank!=0, send the provided error detail to world_rank==0
+     * @param e_detail The error detail to be sent
+     */
+    void sendErrorDetail(AbstractSimRunner::ErrorDetail &e_detail);
+    /**
+     * If world_rank!=0, request a job from world_rank==0 and return the response
+     * @return The index of the assigned job
+     */
+    int requestJob();
+    /**
+     * Wait for all MPI ranks to reach a barrier
+     */
+    void worldBarrier();
+    /**
+     * If world_rank!=0, send the local GPU string to world_rank==0 and return empty string
+     * If world_rank==0, receive GPU strings and assemble the full remote GPU string to be returned
+     */
+    std::string assembleGPUsString();
+    /**
+     * Common function for handling local errors during MPI execution
+     */
+    void retrieveLocalErrorDetail(std::mutex &log_export_queue_mutex,
+        std::multimap<int, AbstractSimRunner::ErrorDetail> &err_detail,
+        std::vector<AbstractSimRunner::ErrorDetail> &err_detail_local, int i);
+
+ private:
+    /**
+     * @return Retrieve the local world rank from MPI
+     */
+    int getWorldRank();
```

Comment:
Any reason these methods are private but underlying variables are public scoped? 

Would make more sense for the mehtods to be pub and the variables to be private.

Reply:
The vars are const, so them being public is fine.

The methods actually trigger MPI to initialise MPI if required before returning the value. They aren't trivial getters, they are initialisers.